feat: add `blas/ext/base/gminheap-sift-down` (#12673)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/gminheap-sift-down/README.md

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# gminheapSiftDown

> > Sift a value down into a strided min-heap.

Reviewer: ```suggestion
> Sift a value down from a specified index in a strided min-heap until the heap property is restored.
```

<section class="usage">

## Usage

```javascript
var gminheapSiftDown = require( '@stdlib/blas/ext/base/gminheap-sift-down' );
```

#### gminheapSiftDown( N, index, value, x, strideX )

> Sifts a value down into a strided min-heap.

Reviewer: ```suggestion
Sifts a value down from a specified index in a strided min-heap until the heap property is restored.
```

```javascript

> var x = [ 7.0, 2.0, 3.0, 4.0, 5.0 ];

Reviewer: ```suggestion
var x = [ 1.0, 2.0, 3.0, 4.0, 5.0 ];
```

gminheapSiftDown( 5, 0, 7.0, x, 1 );
// x => [ 2.0, 4.0, 3.0, 7.0, 5.0 ]
```

The function has the following parameters:

-   **N**: number of indexed elements.
-   **index**: logical index at which to begin sifting.
-   **value**: value to place into the heap.
-   **x**: heap storage array.
-   **strideX**: stride length.

The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sift a value down into a heap occupying every other element:

```javascript

> var x = [ 7.0, 0.0, 2.0, 0.0, 3.0 ];

Reviewer: ```suggestion
var x = [ 1.0, 0.0, 2.0, 0.0, 3.0 ];
```

> The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sift a value down into a heap occupying every other element:

Reviewer: ```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sift a value down a heap occupying every other element:
```

gminheapSiftDown( 3, 0, 7.0, x, 2 );
// x => [ 2.0, 0.0, 7.0, 0.0, 3.0 ]
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial array...

> var x0 = new Float64Array( [ 0.0, 7.0, 2.0, 3.0, 4.0, 5.0 ] );

Reviewer: ```suggestion
var x0 = new Float64Array( [ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0 ] );
```

// Create an offset view...
var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

gminheapSiftDown( 5, 0, 7.0, x1, 1 );
// x0 => <Float64Array>[ 0.0, 2.0, 4.0, 3.0, 7.0, 5.0 ]
```

#### gminheapSiftDown.ndarray( N, index, value, x, strideX, offsetX )

> Sifts a value down into a strided min-heap using alternative indexing semantics.

Reviewer: ```suggestion
Sifts a value down from a specified index in a strided min-heap until the heap property is restored using alternative indexing semantics.
```

```javascript

> var x = [ 7.0, 2.0, 3.0, 4.0, 5.0 ];

Reviewer: ```suggestion
var x = [ 1.0, 2.0, 3.0, 4.0, 5.0 ];
```

gminheapSiftDown.ndarray( 5, 0, 7.0, x, 1, 0 );
// x => [ 2.0, 4.0, 3.0, 7.0, 5.0 ]
```

The function has the following additional parameters:

-   **offsetX**: starting index.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to sift a value down into a heap occupying the last five elements:

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to sift a value down a heap occupying the last five elements:
```

```javascript

> var x = [ 0.0, 7.0, 2.0, 3.0, 4.0, 5.0 ];

Reviewer: ```suggestion
var x = [ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0 ];
```

gminheapSiftDown.ndarray( 5, 0, 7.0, x, 1, 1 );
// x => [ 0.0, 2.0, 4.0, 3.0, 7.0, 5.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If `N <= 0`, both functions return `x` unchanged.
-   The input strided array is modified **in-place** (i.e., the input strided array is **mutated**).
-   Both functions assume that the subtrees rooted at the children of `index` already satisfy the min-heap property and only the value being sifted may violate the min-heap invariant.
-   The min-heap algorithm is sensitive to the presence of `NaN` values. Since `NaN` comparisons always return `false`, if `NaN` values are present in the input array, the results may be unpredictable.
-   Both functions support array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var gminheapSiftDown = require( '@stdlib/blas/ext/base/gminheap-sift-down' );

// Define a min-heap whose root violates the min-heap invariant:
var x = [ 7.0, 2.0, 3.0, 4.0, 5.0 ];
console.log( x );

> // Sift the root value down toward the leaves in order to restore the min-heap invariant:

Reviewer: ```suggestion
// Sift the root value down toward the leaves in order to restore the min-heap property:
```
